Enterprise-CMCS/macpro-mdct-mfp
# WP Copy Over Quarter Logic

### Description

Forms will now copy to the next period correctly with the correct quarters and the status of each entity reflects this change. Only issue is that in the database, it's not smart enough to tell when a form copy for that period was made so it will keep making a new one each time someone else makes one. I pulled this issue into it's own ticket as it'll require looking over the database a bit more.

Will continue adding more coverage in another branch

**IMPORTANT**
For testing locally, add this line of code into create.ts. Each time you copy over a form, it will create the next period's form automatically.
(Don't try to cheat by changing your system's date, it just breaks everything)
```
if (unvalidatedMetadata?.copyReport) {
      Date.prototype.getMonth = () => {
        //if reporting period is 1, we want to generate period 2 but returning a month in the later half of the year
        return unvalidatedMetadata?.copyReport.reportPeriod === 1 ? 11 : 2;
      };
      Date.prototype.getFullYear = () => {
        //if reporting period is 1, we want to return the same year but period 2, else next year
        return unvalidatedMetadata?.copyReport.reportPeriod === 1
          ? unvalidatedMetadata?.copyReport.reportYear
          : (unvalidatedMetadata?.copyReport.reportYear + 1);
      };
    }
```
<img width="954" alt="Screenshot 2023-11-29 at 3 07 25 PM" src="https://github.com/Enterprise-CMCS/macpro-mdct-mfp/assets/127151429/f2aa0ada-79c5-4b8c-8eca-ece6dce3af5c">

In DashboardPage.tsx,, turn off disabled submission check (line 351):
```
    // disabled={isAddSubmissionDisabled()}
```
The reason for this is because it's checking to make sure the system date is not the in the same period as the report in the table

### Related ticket(s)


CMDCT-2863
CMDCT-2339

---
### How to test

1) Add the line above before spinning up your local server 
2) Sign into MFP as a stateuser
3) Fill out a WP and submit
4) Sign out and go into adminuser and approve the WP
5) You should now be able to generate the next period 
6) Sign back into the stateuser, go back to the WP dashboard and click the [Continue Work Plan for Next Period] button
7) A slightly different popup from create the form should show up, listing the form you will copy
8) Copy the previous form by click [Continue from previous period] button
9) Check the forms for any fields that have repeating quarters to be showing an error status

### Important updates




---
### Author checklist


---


_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_
